dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExpressionFactory.java

```diff
@@ -343,6 +344,10 @@ public Expression makeGeneralZero(Type type) {
     }
 
     public Expression makeCast(Expression expression, Type type, boolean signed) {
+        if (expression.getType().equals(type)) {
+            return expression;
+        }
+
```

Comment:
This should probably be done by ExprSimplified` and not the factory (we should not have any magic in here)

Reply:
1. I don't think this is magic, because the user cannot (and should not) rely on the syntactic representation of the returned expression (a reason why all methods should return `Expression` and nothing more specific really). For example, a cast to BooleanType is represented by an ITE expression and is not even a `CastExpression`.
2. ExprSimplifier cannot do this unless we introduce cast expressions for each and every type, in particular unnecessary identity casts like Ptr2Ptr. Therefore, I consider this check to not be an optimization (which ExprSimplifier can do) but rather a choice on how to represent the cast in the first place.